### PR TITLE
feat: add categories API

### LINKED ISSUE #87: Group components in the sidebar
A developer should be able to group components in the left hand sidebar.

### Option 1

Add the `tags` array to the component config object.

Additionally, this can be used restrict which components can be dropped into a DropZone (see #37)

### Option 2

As option 1, but singular `tag`.

### Option 3

A render function may provide similar functionality by allowing the developer to the override the UI. Similar to the fieldset proposal #64.

### PR DESCRIPTION
**Pull request by chrisvxd**
This PR introduces:

* a `categories` API for grouping components
* a `renderComponentList` API for both the user and plugins to enable enhancing of the component list (such as for search)

## `categories` API

`categories` is a new key on `Config`. 

- **[categoryName]** (`object`)
  - **components** (`sting[]`, [optional]): Array containing the names of components in this category
  - **title** (`sting`, [optional]): Title of the category
  - **visible** (`boolean`, [optional]): Whether or not the category should be visible in the side bar
  - **defaultExpanded** (`boolean`, [optional]): Whether or not the category should be expanded in the side bar by default

Example:

```tsx
  categories: {
    layout: {
      components: ["Columns", "Flex", "VerticalSpace"],
    },
    typography: {
      components: ["Heading", "Text"],
    },
    interactive: {
      title: "Actions",
      components: ["ButtonGroup"],
    },
  },
```

This renders a UI like this

image

It's okay for a component to appear in multiple categories.

All remaining components are grouped in a hidden category called `other` (currently undocumented, but worth adding once we deploy the website). By default, the `other` category will show all components that aren't included by other categories, but it can still be configured like any other category.

For example, we might choose to hide it

```tsx
  categories: {
   ...existingCategories
    other: {
      visible: false,
    },
  }
```

image

Or render it collapsed by default

```tsx
  categories: {
   ...existingCategories
    other: {
      defaultExpanded: false,
    },
  }
```

image

### Why this API was chosen

There was much discussion in #87. This proposal implements option 4 from [this comment](https://github.com/measuredco/puck/issues/87#issuecomment-1768303564). There was some interest in option 1b, and this implementation was originally an exploration of both options. However, option 4 was chosen because:

* It's easier to strictly type than 1b, especially when the user splits the components into separate files (as we do in demo)
* You can reorder components
* All categories considerations are kept in one place

## renderComponentList API

Additionally, the `renderComponentList` API was introduced for both the user _and_ plugins, allowing users to add custom UI to the component list. Since the component list is using application state, users can control the component list by writing to the `componentList` UI state object.

For example, here's a search implementation:

https://github.com/measuredco/puck/assets/985961/6b1151af-9b98-4697-8f6a-52ef35170dd0

<details>
<summary>See code</summary>

```tsx
          renderComponentList={({ children, dispatch, state }) => (
            <>
              <input
                placeholder="Search components"
                onChange={(e) => {
                  if (!prevComponentList) {
                    prevComponentList = { ...state.ui.componentList };
                  }

                  if (!e.currentTarget.value && prevComponentList) {
                    dispatch({
                      type: "setUi",
                      ui: {
                        ...state.ui,
                        componentList: prevComponentList,
                      },
                    });

                    prevComponentList = null;

                    return;
                  }

                  const filteredComponentList = Object.keys(
                    config.components
                  ).filter(
                    (componentName) =>
                      componentName
                        .toLowerCase()
                        .indexOf(e.currentTarget.value.toLowerCase()) > -1
                  );

                  dispatch({
                    type: "setUi",
                    ui: {
                      ...state.ui,
                      componentList: {
                        results: {
                          title: `${filteredComponentList.length} result${
                            filteredComponentList.length === 1 ? "" : "s"
                          }`,
                          components: filteredComponentList,
                        },
                        other: { visible: false },
                      },
                    },
                    recordHistory: false,
                  });
                }}
              ></input>

              {children}
            </>
          )}
```

</details>

> **Note, this is the first render API exposed to both. We might want to also consider exposing `renderHeader` and `renderHeaderActions` to plugins at a later date**.

Closes #87

### CODE REVIEW

File: packages/core/components/ComponentList/index.tsx [hunk 1]
```diff
@@ -1,49 +1,110 @@
...
+  title,
+}: {
+  children?: ReactNode;
+  title?: string;
+}) => {
+  const { config } = useAppContext();
+
+  const [isExpanded, setIsExpanded] = useState(true);
+
+  const [counter, setCounter] = useState(0);
```
Comment (chrisvxd) [inline packages/core/components/ComponentList/index.tsx:55]: Remove this

File: apps/demo/app/[...puckPath]/client.tsx [hunk 2]
```diff
@@ -47,6 +48,29 @@ export function Client({ path, isEdit }: { path: string; isEdit: boolean }) {
           }}
           plugins={[headingAnalyzer]}
           headerPath={path}
+          // TODO this feels a bit verbose and clunky
```
Comment (chrisvxd) [inline apps/demo/app/[...puckPath]/client.tsx:51]: Opportunity for more structured tighter API?

**Comment by xaviemirmon** [comment]
@chrisvxd, this looks good! 

A few things I've noticed: 

1) If no categories are selected, the Other accordion is displayed.  If you don't use this new API, it should behave as before with no drop-down.  
2) The "Other" title isn't overrideable as the other titles.  Simple fix, though:
```tsx
<ComponentList id="other" key="other" title={ui.componentList.other?.title ? ui.componentList.other.title : "Other"}>
```
3) This next one is a bugger, but I might have a fix. If you have two of the same components in the sidebar using different categories, the dragged item is the last item because the IDs aren't unique.

```diff
diff --git a/packages/core/components/ComponentList/index.tsx b/packages/core/components/ComponentList/index.tsx
index 133dcc8..154109b 100644
--- a/packages/core/components/ComponentList/index.tsx
+++ b/packages/core/components/ComponentList/index.tsx
@@ -14,15 +14,17 @@ const getClassNameItem = getClassNameFactory("ComponentListItem", styles);
 const ComponentListItem = ({
   component,
   index,
+  category,
 }: {
   component: string;
   index: number;
+  category?: string
 }) => {
   return (
     <div className={getClassNameItem()}>
       <Draggable
         key={component}
-        id={component}
+        id={`${category}-${component}`}
         index={index}
         showShadow
         disableAnimations
diff --git a/packages/core/lib/use-component-list.tsx b/packages/core/lib/use-component-list.tsx
index f3a3189..a739c23 100644
--- a/packages/core/lib/use-component-list.tsx
+++ b/packages/core/lib/use-component-list.tsx
@@ -30,6 +30,7 @@ export const useComponentList = (config: Config, ui: UiState) => {
                   <ComponentList.Item
                     key={componentName}
                     component={componentName as string}
+                    category={categoryKey}
                     index={i}
                   />
                 );
diff --git a/packages/core/reducer/data.ts b/packages/core/reducer/data.ts
index 839da84..de2549f 100644
--- a/packages/core/reducer/data.ts
+++ b/packages/core/reducer/data.ts
@@ -22,10 +22,11 @@ export const addToZoneCache = (key: string, data: Content) => {
 
 export const reduceData = (data: Data, action: PuckAction, config: Config) => {
   if (action.type === "insert") {
+    const [,componentType] = action.componentType.split('-');
     const emptyComponentData = {
-      type: action.componentType,
+      type: componentType,
       props: {
-        ...(config.components[action.componentType].defaultProps || {}),
+        ...(config.components[componentType].defaultProps || {}),
         id: generateId(action.componentType),
       },
     };
```

**Comment by xaviemirmon** [comment]
Nice! @chrisvxd,

3) It's almost there, but the reducer also needs to be updated to strip out the category from the component name:

```diff
diff --git a/packages/core/reducer/data.ts b/packages/core/reducer/data.ts
index 839da84..de2549f 100644
--- a/packages/core/reducer/data.ts
+++ b/packages/core/reducer/data.ts
@@ -22,10 +22,11 @@ export const addToZoneCache = (key: string, data: Content) => {
 
 export const reduceData = (data: Data, action: PuckAction, config: Config) => {
   if (action.type === "insert") {
+    const [,componentType] = action.componentType.split('-');
     const emptyComponentData = {
-      type: action.componentType,
+      type: componentType,
       props: {
-        ...(config.components[action.componentType].defaultProps || {}),
+        ...(config.components[componentType].defaultProps || {}),
         id: generateId(action.componentType),
       },
     };
```

Otherwise, you get this nastiness (btw — thanks for the intro to Kap):

Kapture 2023-10-24 at 13 04 17

**Comment by xaviemirmon** [comment]
All good now @chrisvxd

**Review by monospaced — APPROVED** [review]

**Pull request merged**